src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Instance.hs

```diff
@@ -28,75 +38,85 @@ import           ZkFold.Symbolic.Data.Eq
 
 ------------------------------------- Instances -------------------------------------
 
-instance Arithmetic a => SymbolicData a (ArithmeticCircuit a) where
-    pieces r = [r]
-
-    restore [r] = r
-    restore _   = error "restore ArithmeticCircuit: wrong number of arguments"
-
-    typeSize = 1
-
-instance FiniteField a => Finite (ArithmeticCircuit a) where
-    type Order (ArithmeticCircuit a) = Order a
-
-instance Arithmetic a => AdditiveSemigroup (ArithmeticCircuit a) where
-    r1 + r2 = circuit $ do
-        i <- runCircuit r1
-        j <- runCircuit r2
-        newAssigned (\x -> x i + x j)
-
-instance (Arithmetic a, Scale c a) => Scale c (ArithmeticCircuit a) where
-    scale c r = circuit $ do
-        i <- runCircuit r
-        newAssigned (\x -> (c `scale` one :: a) `scale` x i)
-
-instance Arithmetic a => AdditiveMonoid (ArithmeticCircuit a) where
-    zero = circuit $ newAssigned (const zero)
-
-instance Arithmetic a => AdditiveGroup (ArithmeticCircuit a) where
-    negate r = circuit $ do
-        i <- runCircuit r
-        newAssigned (\x -> negate (x i))
-
-    r1 - r2 = circuit $ do
-        i <- runCircuit r1
-        j <- runCircuit r2
-        newAssigned (\x -> x i - x j)
-
-instance Arithmetic a => MultiplicativeSemigroup (ArithmeticCircuit a) where
-    r1 * r2 = circuit $ do
-        i <- runCircuit r1
-        j <- runCircuit r2
-        newAssigned (\x -> x i * x j)
-
-instance Arithmetic a => Exponent (ArithmeticCircuit a) Natural where
+instance (Arithmetic a, KnownNat n) => SymbolicData a n (ArithmeticCircuit n a) where
+    pieces = id
+
+    restore = ArithmeticCircuit
+
+-- TODO: I had to add these constraints and I don't like them
```

Comment:
This happens because in `Finite`, I specified that `Order a` and `NumberOfBits a` have to be `KnownNat`. Maybe there is a better approach?